src/core/composer/qgscomposerpointsbasedshape.cpp

```diff
@@ -0,0 +1,404 @@
+/***************************************************************************
+                         qgscomposerpointsbasedshape.cpp
+    begin                : March 2016
+    copyright            : (C) 2016 Paul Blottiere, Oslandia
+    email                : paul dot blottiere at oslandia dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgscomposerpointsbasedshape.h"
+#include "qgscomposition.h"
+#include "qgscomposerutils.h"
+#include "qgssymbollayerv2utils.h"
+#include "qgssymbolv2.h"
+#include <limits>
+
+QgsComposerPointsBasedShape::QgsComposerPointsBasedShape( QString tagName,
+    QgsComposition* c )
+    : QgsComposerItem( c ),
+    mTagName( tagName ),
+    mSelectedPoint( -1 ),
+    mDrawPoints( false )
+{
+}
+
+QgsComposerPointsBasedShape::QgsComposerPointsBasedShape( QString tagName,
+    QPolygonF polygon,
+    QgsComposition* c )
+    : QgsComposerItem( c ),
+    mTagName( tagName ),
+    mSelectedPoint( -1 ),
+    mDrawPoints( false )
+{
+  const QRectF boundingRect = polygon.boundingRect();
+  setSceneRect( boundingRect );
+
+  // change polygon coordinate in bounding area reference system
+  const QPointF topLeft = boundingRect.topLeft();
+
+  QVector<QPointF>::iterator it = polygon.begin();
```

Comment:
This could be simplified by using a QTransform, ie

    mPolygon = QTransform::fromTranslate(-topLeft.x(), -topLeft.y() ).map( polygon );